### PR TITLE
Update spec to current explainer

### PR DESCRIPTION
**Pull request by johnivdel**
This updates the draft spec to include all changes which were landed to the explainer since it was last updated.

This includes:
 - all naming changes on #103 and #121
 - removal of hexadecimal encoding for data fields
 - clarification on the value represented by a/attributiondestination

Algorithm definitions and variables are updated to reflect the new naming.

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/conversion-measurement-api/pull/131.html" title="Last updated on Apr 30, 2021, 7:11 PM UTC (32fbcc1)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/conversion-measurement-api/131/6e542cc...32fbcc1.html" title="Last updated on Apr 30, 2021, 7:11 PM UTC (32fbcc1)">Diff</a>

### CODE REVIEW

**Comment by johnivdel** [comment]
@jyasskin could you take a look at this? Thanks!

cc @csharrison

File: index.bs [hunk 1]
```diff
@@ -49,198 +49,207 @@ conversion domain.
 
 # HTML monkeypatches # {#html-monkeypatches}
 
-Rewrite the anchor element to accept the following attributes:
+Rewrite the <{a}> to accept the following attributes:
 
 <pre class="idl">
 partial interface HTMLAnchorElement {
```
Comment (jyasskin) [inline index.bs:55]: Not new in this PR, so better to make another PR to fix it, but this only adds to the `HTMLAnchorElement` interface, and not to the allowed content attributes.

File: index.bs [hunk 2]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
+    [CEReactions, Reflect] attribute DOMString attributionsourceeventid;
+    [CEReactions, Reflect] attribute DOMString attributionreportto;
+    [CEReactions, Reflect] attribute unsigned long long attributionexpiry;
 };
 </pre>
 
-The <dfn for="a" element-attr>conversiondestination</dfn> is the
-declared destination [=scheme-and-registrable-domain=] of the anchor for
-purposes of conversion measurement
+The <dfn for="a" element-attr>attributiondestination</dfn> is an [=url/origin=]
```
Comment (jyasskin) [inline index.bs:63]: Do you define anywhere what should happen if the user sets something that isn't an origin (or isn't even a URL) into this string?

File: index.bs [hunk 3]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
+which declares the intended final navigation url resulting from running
```
Comment (jyasskin) [inline index.bs:64]: An origin won't declare the exact final navigation URL. And we can incorporate the note below into this paragraph. Perhaps

```suggestion
that is intended to be [=same site=] with the origin of the final navigation url resulting from running
```

File: index.bs [hunk 4]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
+Issue: Need monkey patches passing attribution source in navigation, and a mechanism
+for validating the resulting document matches the attributiondestination.
 
 # Structures # {#structures}
 
-<h3 dfn-type=dfn>Impression</h3>
+<h3 dfn-type=dfn>Attribution source</h3>
 
-An impression is a [=struct=] with the following items:
+An attribution source is a [=struct=] with the following items:
```
Comment (jyasskin) [inline index.bs:88]: For a followup, it'd be nice for readers if you include a sentence for each of these structs saying what they're intended to mean. Those sentences are non-normative, so you can wrap them in `<span class=note>` if you're feeling formal, but most spec authors don't bother.

File: index.bs [hunk 5]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
 
-<dl dfn-for="impression">
-: <dfn>impression source</dfn>
+<dl dfn-for="attribution source">
+: <dfn>source origin</dfn>
 :: An [=url/origin=].
```
Comment (jyasskin) [inline index.bs:92]: For a followup, it'd be good to include a statement of what each of the fields is intended to mean, or where it comes from. E.g. this one could be "copied from the clicked <{a}> element's [=Node/node document=]'s [=Document/origin=]."

File: index.bs [hunk 6]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
+    : [=attribution source/expiry=]
+    :: |currentTime| +  <{a/attributionexpiry}> milliseconds.
+    : [=attribution source/source time=]
     :: |currentTime|.
+1. Return |source|
+
+<dfn>Max event id value</dfn> is a vendor specific integer which controls the meximum size value which can be used as an [=attribution source/event id=]
 
-1. Issue: Need to spec how to store the impression.
+1. Issue: Need to spec how to store the attribution source.
```
Comment (jyasskin) [inline index.bs:170]:
```suggestion
Issue: Need to spec how to store the attribution source.
```

File: index.bs [hunk 7]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
+    : [=attribution source/reporting endpoint=]
+    :: The [=url/origin=] of the result of running the [=URL parser=] with
+        |anchor|'s <{a/attributionreportto}> attribute.
+    : [=attribution source/expiry=]
+    :: |currentTime| +  <{a/attributionexpiry}> milliseconds.
+    : [=attribution source/source time=]
     :: |currentTime|.
+1. Return |source|
+
+<dfn>Max event id value</dfn> is a vendor specific integer which controls the meximum size value which can be used as an [=attribution source/event id=]
```
Comment (jyasskin) [inline index.bs:168]:
```suggestion
<dfn>Max event id value</dfn> is a vendor specific integer which controls the maximum size value which can be used as an [=attribution source/event id=]
```

File: index.bs [hunk 8]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
-[=string=] |input| and a unsigned long long |maxData|.
+[=attribution source/event id=] and [=attribution trigger/trigger data=].
 
-<dfn>Parsing conversion data</dfn> from |input| with |maxData| returns the result of
-the following steps:
+To <dfn>parse attribution data</dfn> given a [=string=] |input|, and an unsigned long long
+|maxData| perform the  following steps. They return an unsigned long long:
 
-1. Let |decodedInput| be the result of decoding |input| as a base-16 integer.
+1. Let |decodedInput| be the result of decoding |input| as a base-10 integer.
```
Comment (jyasskin) [inline index.bs:219]: This needs to define what happens if the value isn't a base-10 integer. Probably call https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-non-negative-integers and handle the resulting errors.

File: index.bs [hunk 9]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
+An attribution source is a [=struct=] with the following items:
 
-<dl dfn-for="impression">
-: <dfn>impression source</dfn>
+<dl dfn-for="attribution source">
+: <dfn>source origin</dfn>
 :: An [=url/origin=].
-: <dfn>impression data</dfn>
+: <dfn>event id</dfn>
 :: A [=string=].
```
Comment (jyasskin) [inline index.bs:94]: You're setting this to an integer in [=obtain an attribution source=].

File: index.bs [hunk 10]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
-1. Issue: Need to spec how to store the impression.
+1. Issue: Need to spec how to store the attribution source.
 
-<h3 algorithm id="creating-a-conversion">Creating a conversion</h3>
+<h3 algorithm id="attribution-trigger-creation">Creating an attribution trigger</h3>
 
-To <dfn>create a conversion</dfn> from a [=url=] |url| and an
-[=environment settings object=] |environment|, return a new [=conversion=]
-struct with the items:
+To <dfn>obtain an attribution trigger</dfn> given a [=url=] |url|, and an
```
Comment (jyasskin) [inline index.bs:174]:
```suggestion
To <dfn>obtain an attribution trigger</dfn> given a [=url=] |url| and an
```

File: index.bs [hunk 11]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
-    :: The result of running [=parse a conversion destination=] on |anchor|.
-    : [=impression/reporting endpoint=]
-    :: The [=url/origin=] of the result of running the [=URL parser=] on the value
-        of |anchor|'s <{a/reportingorigin}> attribute.
-    : [=impression/expiry=]
-    :: |currentTime| +  <{a/impressionexpiry}> milliseconds.
-    : [=impression/impression time=]
+    : [=attribution source/event id=]
+    :: The result of running [=parse attribution data=] with |anchor|'s
+        <{a/attributionsourceeventid}> attribute and [=max event id value=].
```
Comment (jyasskin) [inline index.bs:156]: Here I'd probably phrase it 
```suggestion
        <{a/attributionsourceeventid}> attribute modulo [=max event id value=].
```

and match that in the definition. It doesn't quite match the advice in https://infra.spec.whatwg.org/#algorithm-params, but I think it winds up more readable.

File: index.bs [hunk 12]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
+    : [=attribution source/attribution destination=]
+    :: The result of running [=parse an attribution destination=] with |anchor|'s <{a/attributiondestination}> attribute.
+    : [=attribution source/reporting endpoint=]
+    :: The [=url/origin=] of the result of running the [=URL parser=] with
```
Comment (jyasskin) [inline index.bs:160]: The URL parser can return an error, and you need to say what happens in that case.

File: index.bs [hunk 13]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
-To <dfn>create a conversion</dfn> from a [=url=] |url| and an
-[=environment settings object=] |environment|, return a new [=conversion=]
-struct with the items:
+To <dfn>obtain an attribution trigger</dfn> given a [=url=] |url|, and an
+[=environment settings object=] |environment|, return a [=attribution trigger=]
+with the items:
 
-    : [=conversion/conversion source=]
+    : [=attribution trigger/trigger origin=]
     :: |environment|'s [=environment/top-level origin=].
```
Comment (jyasskin) [inline index.bs:179]: Why is this the top-level origin instead of just the origin? Can attribution triggers happen in cross-origin iframes? Have you discussed somewhere the security implications of writing a different origin into the attribution trigger, and what it means for specs that need to trust or distrust that value?

File: index.bs [hunk 14]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
+1. If |request|'s [=request/current url's=] [=url/path=] is not `.well-known/attribution-reporting/trigger-attribution`,
     return.
 1. If |request|'s [=request/redirect count=] is less than 1, return.
 1. Let |previousUrl| be the second to last [=URL=] in |request|'s
     [=request/URL list=].
 1. If |request|'s [=request/current url's=] [=url/origin=] is not [=same origin=] with
     |previousUrl|'s [=url/origin=], return.
-1. Let |conversionToRegister| be the result of applying [=create a conversion=] from the
-    request's [=request/current url=].
+1. Let |trigger| be the result of running [=obtain an attribution trigger=] with
```
Comment (jyasskin) [inline index.bs:203]: This doesn't pass the environment settings object you expect above.

**Review by jyasskin — CHANGES_REQUESTED** [review]
I haven't gotten all the way through the spec, but here's a bunch of comments to start on.

File: index.bs (hunk 4 above)
Comment (johnivdel) [reply index.bs:84]: Ack.

File: index.bs (hunk 5 above)
Comment (johnivdel) [reply index.bs:88]: Ack.

File: index.bs (hunk 9 above)
Comment (johnivdel) [reply index.bs]: Fixed. I noted this is a 64-bit integer, but I couldn't find any precedence in the html spec for this.

File: index.bs [hunk 15]
```diff
@@ -49,198 +49,207 @@ conversion domain.
...
 <pre class="idl">
 partial interface HTMLAnchorElement {
-    [CEReactions, Reflect] attribute DOMString conversiondestination;
-    [CEReactions, Reflect] attribute DOMString impressiondata;
-    [CEReactions, Reflect] attribute DOMString reportingorigin;
-    [CEReactions, Reflect] attribute unsigned long long impressionexpiry;
+    [CEReactions, Reflect] attribute DOMString attributiondestination;
+    [CEReactions, Reflect] attribute DOMString attributionsourceeventid;
+    [CEReactions, Reflect] attribute DOMString attributionreportto;
+    [CEReactions, Reflect] attribute unsigned long long attributionexpiry;
```
Comment (johnivdel) [inline index.bs:59]: unsigned long long is not a type defined in https://html.spec.whatwg.org/#reflecting-content-attributes-in-idl-attributes

This either needs to be an unsigned long or a string. I can address this in a separate PR.

File: index.bs (hunk 1 above)
Comment (johnivdel) [reply index.bs:55]: Ack.

File: index.bs [hunk 16]
```diff
@@ -1,13 +1,13 @@
 <pre class='metadata'>
-Title: Conversion Measurement
-Shortname: conversion-measurement
+Title: Attribution Reporting
+Shortname: attribution-reporting
```
[the remark this answers is not in the document]
Comment (johnivdel) [reply index.bs:3]: Hopefully one change is enough :)

**Review by johnivdel — COMMENTED** [review]
Thanks for the great comments Jeffrey!

File: index.bs [hunk 17]
```diff
@@ -49,198 +49,223 @@ conversion domain.
...
+1. Let |sourceOrigin| be |anchor|'s [=relevant settings object=]'s
+    [=environment/top-level origin=].
+1. Let |reportingOrigin| be |sourceOrigin|.
+1. If |anchor| has an <{a/attributionreportto}> attribute, then:
+    1. Let |reportingUrl| be the result of running the
+        [=URL parser=] with |anchor|'s <{a/attributionreportto}> value
+    1. If |reportingUrl| is failure or null, return null.
+    1. Set |reportingOrigin| to |reportingUrl|'s [=url/origin=].
+1. If |anchor| has an <{a/attributionexpiry}> attribute, let |expiry| be that 
+    attribute's value in milliseconds.
```
Comment (jyasskin) [inline index.bs:162]: Use HTML's integer parsing rules here too.

File: index.bs [hunk 18]
```diff
@@ -49,198 +49,223 @@ conversion domain.
...
+    :: |sourceOrigin|
+    : [=attribution source/event id=]
+    :: The result of running [=parse attribution data=] with |anchor|'s
+        <{a/attributionsourceeventid}> attribute modulo [=max event id value=].
+    : [=attribution source/attribution destination=]
+    :: |attributionDestination|
+    : [=attribution source/reporting endpoint=]
+    :: |reportingOrigin|
+    : [=attribution source/expiry=]
+    :: |currentTime| +  |expiry|
```
Comment (jyasskin) [inline index.bs:177]:
```suggestion
    :: |currentTime| + |expiry|
```

File: index.bs (hunk 13 above)
Comment (jyasskin) [reply index.bs:194]: I was worried about a hostile iframe being able to create somehow-hostile attributions to the surrounding page, but I'm not sure what attacks that would actually be useful for. It helps that the user has to click in order to make the attribution source happen. Maybe hostile triggers are more worrying, since they might cause payment from the trigger to the source? In any case, this sort of attribution (no pun intended) of one origin's actions to a different origin is worth talking about in the security considerations.

**Review by jyasskin — APPROVED** [review]

File: index.bs (hunk 13 above)
Comment (johnivdel) [reply index.bs:194]: Yes that would cause issues. Sources are also problematic: a hostile frame logging additional sources could cause credits to be assigned incorrectly, or cause the browser to delete already existing sources in some cases.

The main mitigation we have is [requiring a permissions policy](https://github.com/WICG/conversion-measurement-api#publisher-side-controls-for-attribution-source-declaration) to invoke the API inside of iframes, which is disabled by default for cross-origin iframes.

I will make a note to address this in Security Considerations when the Permissions Policy is added to this draft spec.
